jax_distribution/systems/ff_mappo_rware.py

```diff
@@ -0,0 +1,523 @@
+    learn = jax.pmap(learn, axis_name="device")
+
+    # Broadcast params and optimiser state to cores and batch.
+    broadcast = lambda x: jnp.broadcast_to(
+        x, (n_devices, config["update_batch_size"]) + x.shape
+    )
+    params = jax.tree_map(broadcast, params)
+    opt_state = jax.tree_map(broadcast, opt_state)
+
+    # Initialise environment states and timesteps.
+    rng, *env_rngs = jax.random.split(
+        rng, n_devices * config["update_batch_size"] * config["num_envs"] + 1
+    )
+    env_states, timesteps = jax.vmap(env.reset, in_axes=(0))(
+        jnp.stack(env_rngs),
+    )
+
+    # Split rngs for each core.
+    rng, *step_rngs = jax.random.split(rng, n_devices * config["update_batch_size"] + 1)
+    # Add dimension to pmap over.
+    reshape_step_rngs = lambda x: x.reshape(
+        (n_devices, config["update_batch_size"]) + x.shape[1:]
+    )
+    step_rngs = reshape_step_rngs(jnp.stack(step_rngs))
+    reshape_states = lambda x: x.reshape(
+        (n_devices, config["update_batch_size"], config["num_envs"]) + x.shape[1:]
+    )
+    env_states = jax.tree_util.tree_map(reshape_states, env_states)
+    timesteps = jax.tree_util.tree_map(reshape_states, timesteps)
+
+    init_learner_state = LearnerState(
+        params, opt_state, step_rngs, env_states, timesteps
+    )
+    return learn, network, init_learner_state
+
+
+def run_experiment(_run: run.Run, _config: Dict, _log: SacredLogger) -> None:
+    """Runs experiment."""
+    # Logger setup
+    config = config_copy(_config)
+    log = logger_setup(_run, config, _log)
+
+    generator = RandomGenerator(**config["rware_scenario"])
+    # Create envs
+    env = jumanji.make(config["env_name"], generator=generator)
+    env = RwareMultiAgentWithGlobalStateWrapper(env)
+    env = AutoResetWrapper(env)
+    env = LogWrapper(env)
+    eval_env = jumanji.make(config["env_name"], generator=generator)
+    eval_env = RwareMultiAgentWithGlobalStateWrapper(eval_env)
+
+    # PRNG keys.
+    rng, rng_e, rng_p = jax.random.split(jax.random.PRNGKey(config["seed"]), num=3)
+    # Setup learner.
+    learn, network, learner_state = learner_setup(env, (rng, rng_p), config)
+
+    # Setup evaluator.
+    evaluator, absolute_metric_evaluator, (trained_params, eval_rngs) = evaluator_setup(
+        eval_env, rng_e, network, learner_state.params, config
+    )
+
+    # Calculate total timesteps.
+    n_devices = len(jax.devices())
+    config["num_updates_per_eval"] = config["num_updates"] // config["num_evaluation"]
+    timesteps_per_training = (
+        n_devices
+        * config["num_updates_per_eval"]
+        * config["rollout_length"]
+        * config["update_batch_size"]
+        * config["num_envs"]
+    )
+
+    # Run experiment for a total number of evaluations.
+    max_episode_return = jnp.float32(0.0)
+    best_params = None
+    for i in range(config["num_evaluation"]):
+        # Train.
+        with TimeIt(
```

Comment:
Quite nice to have the compilation time be reported on the fly like this!

Reply:
+1, very cool